ocaml/typing/solver.ml

```diff
@@ -484,88 +485,70 @@ module Solver_mono (C : Lattices_mono) = struct
     let vlower = Option.value vlower ~default:[] in
     { upper; lower; vlower; id }
 
-  let submode_try (type a r l) (obj : a C.obj) (a : (a, allowed * r) mode)
-      (b : (a, l * allowed) mode) =
-    let log = Some (ref []) in
+  let submode (type a r l) (obj : a C.obj) (a : (a, allowed * r) mode)
+      (b : (a, l * allowed) mode) ~log =
     let submode_cc ~log:_ obj left right =
       if C.le obj left right then Ok () else Error { left; right }
     in
     let submode_mvc ~log obj v right =
       Result.map_error
         (fun left -> { left; right })
-        (submode_mvc ~log obj v right)
+        (submode_mvc ~log:(Some log) obj v right)
```

Comment:
I'm a little worried about how often we might allocate these `Some` nodes. Maybe it makes sense to have `submode` take an optional log just to avoid the repeated allocation? Not sure.

Reply:
Good point about the allocation. I lifted the allocation to the top level of `submode`, which means just one allocation per `Solver.submode`. Note that `Mode.Value.submode` consists of two `Solver.submode`; so if we lift the allocation to `Mode.Value.submode` we can reduce the allocation further.

That would mean `solver.submode` as you said takes optional log. But I think solver not logging is just wrong and we would be exposing the unsafe capability that the user should never use.

Reply:
Well on a second thought, considering that `mode.ml` is the only user of `solver.mli`, both of which are maintained by the same set of people, it might be fine to allow `solver.mli` be a bit "primitive" and expose bare metals.